src/Storages/StorageAggregatingMemory.cpp

```diff
@@ -0,0 +1,513 @@
+    if (aggregator_transform->params.keys_size == 0 && !aggregator_transform->params.empty_result_for_aggregation_by_empty_set)
+    {
+        AggregatingOutputStream os(*this, getInMemoryMetadataPtr(), context);
+        os.write(src_block_header);
+    }
+}
+
+void StorageAggregatingMemory::startup()
+{
+    try
+    {
+        lazyInit();
+    }
+    catch (Exception & e)
+    {
+        e.addMessage("Failed to initialize AggregatingMemory on startup");
+        LOG_ERROR(log, "{}", getCurrentExceptionMessage(true));
+    }
+}
+
+Pipe StorageAggregatingMemory::read(
+    const Names & column_names,
+    const StorageMetadataPtr & metadata_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t /*max_block_size*/,
+    unsigned num_streams)
+{
+    lazyInit();
+    metadata_snapshot->check(column_names, getVirtuals(), getStorageID());
+    // TODO implement O(1) read by aggregation key
+
+    auto prepared_data = aggregator_transform->aggregator.prepareVariantsToMerge(many_data->variants);
+    auto prepared_data_ptr = std::make_shared<ManyAggregatedDataVariants>(std::move(prepared_data));
+
+    ProcessorPtr source;
+
+    auto filter_key = getFilterKeys(aggregator_transform->params, query_info);
+    source = std::make_shared<ConvertingAggregatedToChunksTransform>(aggregator_transform, std::move(prepared_data_ptr), num_streams);
+
+    StoragePtr mergable_storage = StorageSource::create(source_storage->getStorageID(), source_storage->getInMemoryMetadataPtr()->getColumns(), source);
+
+    ContextMutablePtr local_context = Context::createCopy(context);
+    local_context->addViewSource(mergable_storage);
+
+    InterpreterSelectQuery select(metadata_snapshot->getSelectQuery().inner_query, local_context, SelectQueryOptions());
+    BlockIO select_result = select.execute();
+
+    return QueryPipeline::getPipe(std::move(select_result.pipeline));
+}
+
+BlockOutputStreamPtr StorageAggregatingMemory::write(const ASTPtr & /*query*/, const StorageMetadataPtr & metadata_snapshot, ContextPtr context)
+{
+    lazyInit();
+
+    auto out = std::make_shared<AggregatingOutputStream>(*this, metadata_snapshot, context);
+    return out;
+}
+
+void StorageAggregatingMemory::drop()
+{
+    if (!is_initialized)
+        return;
+
+    /// Drop aggregation state.
+    many_data = nullptr;
+}
+
+void StorageAggregatingMemory::truncate(const ASTPtr &, const StorageMetadataPtr &, ContextPtr context, TableExclusiveLockHolder &)
+{
+    lazyInit();
+
+    /// Assign fresh state.
+    initState(context);
+}
+
+std::optional<UInt64> StorageAggregatingMemory::totalRows(const Settings &) const
+{
+    if (!is_initialized)
+        return 0;
+
+    return many_data->variants[0]->size();
+}
+
+std::optional<UInt64> StorageAggregatingMemory::totalBytes(const Settings &) const
+{
+    if (!is_initialized)
+        return 0;
+
+    // Not possible to determine.
```

Comment:
Actually, we can estimate it (if add something like `many_data->variants[0]->size()` manually).
Generally, it is a size of hash table (`FixedHashTable::getBufferSizeInBytes`) + size of arenas.

Maybe we will add it later.